pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/CompoundDataBuffer.java

```diff
@@ -0,0 +1,672 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.spi.memory;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.BufferUnderflowException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.ArrayList;
+import java.util.List;
+import org.roaringbitmap.buffer.ImmutableRoaringBitmap;
+
+
+/**
+ * A {@link DataBuffer} that is composed of multiple {@link DataBuffer}s that define a single contiguous buffer.
+ * <p>
+ * While reads and writes can span multiple buffers, there may be a performance impact when doing so.
+ * Therefore it is recommended to try to wrap independent buffers.
+ * <p>
+ * Once this class is built, buffers cannot be added or removed.
+ */
+public class CompoundDataBuffer implements DataBuffer {
+
+  private final DataBuffer[] _buffers;
+  private final long[] _bufferOffsets;
+  private int _lastBufferIndex = 0;
+  private final ByteOrder _order;
+  private final long _size;
+  private final boolean _owner;
+
+  /**
+   * Creates a compound buffer from the given buffers.
+   *
+   * @param buffers The buffers that will be concatenated to form the compound buffer.
+   * @param order The byte order of the buffer. Buffers in the array that have a different byte order will be converted.
+   * @param owner Whether this buffer owns the underlying buffers. If true, the underlying buffers will be released when
+   *              this buffer is closed.
+   */
+  public CompoundDataBuffer(DataBuffer[] buffers, ByteOrder order, boolean owner) {
+    _owner = owner;
+    _buffers = buffers;
+    _bufferOffsets = new long[buffers.length];
+    _order = order;
+    long offset = 0;
+    for (int i = 0; i < buffers.length; i++) {
```

Comment:
We can combine the 2 for loops. Cache `buffers[i].size()` which is used multiple times

Reply:
I'm splitting it in two loops on purpose. The idea (not backed by data) is that by doing so the JIT will have an easier time to optimize the second loop:
```
    for (int i = 0; i < buffers.length; i++) {
      _bufferOffsets[i] = offset;
      long size = buffers[i].size();
      offset += size;
    }
```

TBH I don't think one or the other alternative is going to be specially faster, but I think this way is easier to read. The first loop is checking preconditions while the second initialize a couple of complex variables (`_bufferOffsets` and `_size`)